File: templates/problems/12_advanced_graphs/01_reconstruct_itinerary/python/solution.py
def solve_find_itinerary(tickets: list[list[str]]) -> list[str]:
    """Reconstruct the itinerary using Hierholzer's algorithm.

    Time: O(E log E) for sorting, Space: O(E).
    """
    graph: dict[str, list[str]] = {}
    for t in tickets:
        graph.setdefault(t[0], []).append(t[1])
    for k in graph:
        graph[k].sort()

    result: list[str] = []

    def dfs(airport: str) -> None:
        neighbors = graph.get(airport, [])
        while neighbors:
            nxt = neighbors.pop(0)
            dfs(nxt)
        result.append(airport)

    dfs("JFK")
    result.reverse()
    return result

File: templates/problems/12_advanced_graphs/01_reconstruct_itinerary/typescript/solution.ts
/**
 * Reconstruct the itinerary using Hierholzer's algorithm.
 * Time: O(E log E) for sorting, Space: O(E).
 */
export function solveFindItinerary(tickets: string[][]): string[] {
    const graph = new Map<string, string[]>();
    for (const t of tickets) {
        if (!graph.has(t[0])) graph.set(t[0], []);
        graph.get(t[0])!.push(t[1]);
    }
    for (const k of graph.keys()) {
        graph.get(k)!.sort();
    }

    const result: string[] = [];

    const dfs = (airport: string): void => {
        const neighbors = graph.get(airport);
        while (neighbors && neighbors.length > 0) {
            const next = neighbors.shift()!;
            dfs(next);
        }
        result.push(airport);
    };

    dfs("JFK");
    result.reverse();
    return result;
}
